[docs] Drop IE 11 official support

iammminzzy opened the pull request:
This is the first step towards #14420. This PR official communicates that IE 11 is no longer supported. This is a breaking change.

More changes will be required to close the issue, but they won't be breaking changes in theory as IE 11 is no longer supported https://github.com/mui/material-ui/pull/41611#issuecomment-2027080816.

ZeeshanTamboli reviewed (commented):
@iammminzzy Thank you for the pull request. Apart from the code updates, we should also remove the `legacy` bundle build, as discussed in #41568 (refer to https://github.com/mui/material-ui/pull/41568#issuecomment-2008633724). Additionally, we need to remove `ie 11` from the `.browserslistrc` file. Perhaps we should also update [the docs](https://mui.com/material-ui/getting-started/supported-platforms/#ie-11).

ZeeshanTamboli commented:
> > @iammminzzy Thank you for the pull request. Apart from the code updates, we should also remove the `legacy` bundle build, as discussed in #41568 (refer to [#41568 (comment)](https://github.com/mui/material-ui/pull/41568#issuecomment-2008633724)). Additionally, we need to remove `ie 11` from the `.browserslistrc` file. Perhaps we should also update [the docs](https://mui.com/material-ui/getting-started/supported-platforms/#ie-11).
> 
> Thank you for your advice. Should I push these changes to this PR or can this PR be landed as an incremental improvement, since there are still many more things to handle to remove IE 11 support completely?

This should be done in a single PR; otherwise, it might remain incomplete if not addressed in separate PRs.

iammminzzy commented:
@ZeeshanTamboli I followed the approach in https://github.com/mui/material-ui/pull/41568 and removed legacy bundle and updated the docs. Feel free to push any further updates to this PR.

ZeeshanTamboli reviewed (commented):
> There were many style-related references but I thought they were riskier to change, so I left those alone.

Could you try to replace them?

> There was also https://github.com/mui/material-ui/blob/0102a9579628d48d784511a562b7b72f0f51847e/packages/mui-material/src/Tabs/Tabs.js#L423-L425 but changing `isNaN` to `Number.isNaN` seemed wrong despite what the comment says (I believe `typeof indicatorStyle[startIndicator] === 'number'` should be used).

Why `Number.isNaN` seems wrong from `typeof number`?

> Additionally, I wasn't sure if `stableSort()` was safe to drop because I'm not sure if 2020 browsers are meant to be supported: https://github.com/mui/material-ui/blob/0102a9579628d48d784511a562b7b72f0f51847e/docs/src/components/showcase/FolderTable.tsx#L48-L52 

I believe we can safely remove it. It's also included in a demo, and we are supporting browsers that fully support `Array.prototype.sort()`.

NMinhNguyen commented:
@ZeeshanTamboli I’m curious as to why the scope of this PR has to keep increasing, when the author evidently intended to only help clean up some JS references. The original PR that updated `browserslist` also didn’t try to solve the whole old browser support issue and was landed in an incremental fashion. Surely all these other issues can be addressed in follow-up PRs by a core member?

ZeeshanTamboli commented:
> @ZeeshanTamboli I’m curious as to why the scope of this PR has to keep increasing, when the author evidently intended to only help clean up some JS references. 

@NMinhNguyen I believe we shouldn't remove the old code references until the browser support has been officially removed. If we ship the replaced code without actually removing the browser support, it could lead to potential issues. What's the guarantee that the removal of IE 11 browser will be in the same release as this? We need to ensure that the removal of IE 11 browser support coincides with this change.

> The original PR that updated `browserslist` also didn’t try to solve the whole old browser support issue and was landed in an incremental fashion.

It only updated the supported browsers list because the removal of IE 11 could be done separately. In my opinion, they were not interconnected.

NMinhNguyen commented:
> What's the guarantee that the removal of IE 11 browser will be in the same release as this?

There is a stated goal of removing IE11 support as part of v6, and it's relatively recent that the repo has switched from `master` to `next` as the main branch, so I would say that there is more than enough time for this work to be picked up subsequently, without needing to block this contribution.

oliviertassinari commented:
> Surely all these other issues can be addressed in follow-up PRs by a core member?

The smallest possible scope of this PR is to update all docs to not mention IE 11 anymore. Closing #14420 requires updating all the code. So this could be one or many PRs, whatever is smoother.

NMinhNguyen commented:
> > Surely all these other issues can be addressed in follow-up PRs by a core member?
> 
> The smallest possible scope of this PR is to update all docs to not mention IE 11 anymore. It's like a 10 lines removal.
> 
> Closing #14420 requires updating all the code. So this could be one or many PRs, whatever is smoother.

Sounds good, I think this PR should definitely have a limited scope and not close the linked issue, merely reference, which is what it did before @ZeeshanTamboli ’s edit.

Let’s split out the JS reference changes into a separate PR that also doesn’t close the linked issue, merely reference.

Finally, once all the IE references are removed (CSS etc.), then the issue can be closed in that PR.

NMinhNguyen commented on docs/data/material/getting-started/supported-platforms/supported-platforms.md:
```diff
@@ -23,17 +23,8 @@ You can expect Material UI's components to render without major issues.
 
 ### IE 11
```
Should IE be removed from the table above as well? MDN docs no longer mention IE, for example.

iammminzzy replied:
Removed the header and column as well.

iammminzzy commented:
Thanks for your input, everyone!
I have removed all IE references from `.md` files and reverted all other changes.

Please let me know if this is in a mergeable state now.

iammminzzy commented on babel.config.js:
```diff
@@ -153,12 +153,6 @@ module.exports = function getBabelConfig(api) {
-      legacy: {
-        plugins: [
-          // IE11 support
-          '@babel/plugin-transform-object-assign',
```
Out of scope for this PR.

oliviertassinari commented on docs/data/material/migration/migration-v4/migration-v4.md:
```diff
@@ -67,7 +67,7 @@ The default bundle supports the following minimum versions:
 
 Material UI no longer supports IE 11.
-If you need to support IE 11, check out our [legacy bundle](/material-ui/guides/minimizing-bundle-size/#legacy-bundle).
+If you need to support IE 11, check out our [legacy bundle](https://v5.mui.com/material-ui/guides/minimizing-bundle-size/#legacy-bundle).
```
```suggestion
If you need to support IE 11, check out the [legacy bundle](https://v5.mui.com/material-ui/guides/minimizing-bundle-size/#legacy-bundle).
```

ZeeshanTamboli reviewed (commented):
I still don't fully agree with the decision to separate it, but if that's what's been decided, then I'm okay with it. As a community contributor rather than a team member, my suggestions may not carry much weight. Anyway you have already created another PR to drop the code: https://github.com/mui/material-ui/pull/41709, so it's fine. 👍 

I've identified two other places where we need to remove it:

1. From the `stickyHeader` prop of the Table component. You can find it here: [Table Prop - stickyHeader](https://mui.com/material-ui/api/table/#table-prop-stickyHeader). Once text is removed from the type definition, remember to run `pnpm proptypes && pnpm docs:api` to update the Proptypes and API docs.

2. Not directly related to Material UI, but we can also remove it from Joy UI:
   - [Table Prop - stickyFooter](https://mui.com/joy-ui/api/table/#table-prop-stickyFooter)
   - [Table Prop - stickyHeader](https://mui.com/joy-ui/api/table/#table-prop-stickyHeader)

NMinhNguyen commented:
@ZeeshanTamboli thanks for the pointers, I believe the API docs have now been updated too, so there shouldn't be any remaining references to IE in the v6 docs.

ZeeshanTamboli reviewed (approved):
Looks good to me. Since it is a breaking change, I'd like to have others review it as well before merging.

Janpot reviewed (approved):
👍 I see no problem with splitting this.
